### PR TITLE
Add image optimization

### LINKED ISSUE #26: Extend imaging module to handle optimization
Image optimization is something that's one of the most common tasks done by a scraper. We can extend the imaging module to have this incorporated in a well-tested manner in scraperlib and have support for various formats so that we do not face issues with this in individual scrapers. We can have something similar to the video module where we have presets and can even manage different optimizer versions. Maybe we can experiment  a bit here (since we plan to start having formats like WebP in scrapers. See https://github.com/openzim/sotoki/issues/151)

### PR DESCRIPTION
**Pull request by satyamtg**
This adds image optimization to the `imaging` module and breaks it up into several parts inside a submodule called `image`. This basically aims to fix #26 
The following changes are made -
- `zimscraperlib.imaging` is broken into `convertion.py`, `probing.py` and `transsform.py`  and moved to `zimscraperlib.image`. These new files contain the following functions -
    * convertion.py - `convert_image` and `create_favicon`
    * probing.py - `get_colors` and `is_hex_color` (maybe we shall remove this)
    * transform.py - resize_image
- The save_image from imaging module has been moved to `__init__.py` in the image module
- The `ALPHA_NOT_SUPPORTED` constant is moved to zimscraperlib.constants.py
- optimization.py is introduced inside the newely created image module. It is as follows -
    * It contains a class ImageOptimizer which contains the optimization methods for several image types and holds the image optimization options as a dict, such as png_options, jpeg_options etc. (PNG and JPEG supported as of now)
    * ImageOptimizer also contains a method optimize_image to optimize image by detecting its filetype. It also supports overriding the main optimization parameters set at the time of object creation
    * The current PNG and JPEG optimizations are performed using [this](https://github.com/victordomingos/optimize-images) that does it in pure python
    * As far as PNG compression goes, in general we can achieve better compression than what we do with sotoki at the cost of a little color reduction (but then pngquant also reduces color). On the commons.png in this very repository, pngquant + advdef made it from 10KB to around 5.5KB, whereas this optimizer did like 3.9KB.
    * As far as JPEG compression is concerned, it performs little worse than the jpegoptim thing that we used. On the square.jpg in this repository, jpegoptim made it from 130KB to 28KB, whereas this one did it to 35KB. All this was performed with JPEG quality set to 50 (the same as set in our jpegoptim call).

Some refactor and other changes would be needed here (that's why WIP) and we might also want to support WebP. What do you think @rgaudin ?

### CODE REVIEW

**Comment by satyamtg** [comment]
I have added optimization support for WebP and GIF images, tests for them and some presets. The WebP optimization takes place using Pillow whereas the GIF optimization makes use of gifsicle. I first tried Pillow based pure python optimization for GIF but soon realized that it was a particularly bad idea to optimize GIFs using Pillow as even after trying a lowest settings possible, I was only able to get a 18 KB GIF optimized to a 56 KB GIF. Similar things happened with many other GIFs and I was at no point able to get a size below what the GIF was originally made for any GIF I downloaded from the internet which wasn't made by Pillow itself. 

gifsicle is quite good with GIFs and with the use of the --lossy option in version 1.92, we can get even smaller GIFs that what we currently do in Sotoki.

**Comment by rgaudin** [comment]
Please rebase off master

**Comment by satyamtg** [comment]
> Please rebase off master

@rgaudin did it.

**Review by rgaudin — CHANGES_REQUESTED** [review]
Thank you for this complete PR! Still have some concerns but **it's mostly refactoring**.

- Please mention gifsicle dependency in README as well.
- As I've mentioned already, I don't like this `Optimization` class pattern.
I can't see the value it brings but I do see the drawbacks:
  - we have to initialize it. Why do we need to initialize something to optim an image?
  - it's not coherent with the rest of the API which is solely function based.
If for some reason, it's required to, we surely can change it but at the moment I don't see why.
The ImageOptimizer actually collects nothing but default options for each format… reinforcing the idea that we're trying to bundle together stuff that doesn't need to. And there's no setup at all.
- [jpg_png] the `tmp_path` is only deleted upon success.
- I think combining png and jpeg in the same method is not necessary. Those do share some code but it's not that much and some can probably be shared in some private function.
- I think it'd be better to have defaults clearly exposed. We can't make much assumptions on the usage of this tool so the defaults should be the least destructive possible and easily discoverable/over-writable. `max_colors=256` for JPEG is quite scary in that sense.
- I don't think we should print any log on this.
- Presets are not tested (we test content but don't use them)
- As for the API, here are my suggestions:
  - one function for each format
  - default (no option) should be max quality/lossless
  - we should offer 3 pressets (Low, Medium, High). Your two are too agressive. Lots of usage require good quality but still need compression.
  - I don't like the options dict. Doesn't feel natural. We could use some `**options` keyword arguments but given we currently don't use any unexpected value, I'd just set all arguments explicitly and add an (unused for now) **kwargs for future-proofness.
  - those arguments should be ordered by expected usage.
  - we need to move the args values/examples to the function's docstrings
  - I understand `delete_src` is only used if src is different to dst but still I think default should be to not delete something we didn't create ourselves. So defaulting to False.
- we should have a quick check in each method for match between source format and the optim function
Example `optimize_webp` signature

```py
def ensure_matches(src: pathlib.Path, fmt: str):
    """ Raise ValueError if src is not of image type `fmt` """
    pass


def optimize_webp(
    src: pathlib.Path,
    dst: pathlib.Path,
    delete_src: Optional[bool] = False,
    lossless: bool = True,
    quality: int = 50,
    method: int = 3,
    **kwargs):
    """ Reencode WebP image with options as:

        lossless: Whether to use lossless compression (boolean)
            values: True | False
        quality: WebP quality for lossy
                 Effort put into compression for lossless (integer between 0 to 100)
            values: 30 | 45 | 100 | XX
        method: Quality/speed trade-off
                higher values give better compression (integer between 1 and 6)
            values: 1 | 2 | 3 | 4 | 5 | 6 """
    ensure_matches(src, "WEBP")
    pass
```

In addition to this explicit API, we could have another endpoint that abstracts the format selection.

```py
def optimize_image(src, dst, delete_src=False, **options):
    func = {
        "JPEG": optimize_jpeg,
        "PNG": optimize_png,
        "GIF", optimize_gif,
        "WEBP": optimize_webp}.get(format_for(src))
     return func(src, dst, **options)

optimize_image(src, dst, **{**WebpHigh().options, **GifMedium().options})
```

One decision we have to make would be whether we allow converting in the process. As I wrote above, I think those functions should not but we could have an additional one (maybe at `image` module) that combines conversion (if different) and optimization. Let me know what you think about that.

**Comment by satyamtg** [comment]
> One decision we have to make would be whether we allow converting in the process. As I wrote above, I think those functions should not but we could have an additional one (maybe at `image` module) that combines conversion (if different) and optimization. Let me know what you think about that.

For the convertion thing, I think having a small function exposed at the image module to handle convertion before optimizing is a good idea. But shall we have a explicit argument say allow_convert=True or convert=PNG for instance, or should we check the input and output format and do that implicitly? I think we can go with allow_convert=True and then raise an error if convertion is required but not allowed. We would in that case check the format to convert to using the dst suffix. What do you think?

**Comment by rgaudin** [comment]
I think we could do as we do for conversion:
- we find out `src` format from pillow
- we find out `dst` format from `fmt` param if specified or from `dst` filename.
- fail if we can't guess it obviously
- if different, then convert
- call optimization

**Comment by satyamtg** [comment]
> I think we could do as we do for conversion:
> 
> * we find out `src` format from pillow
> * we find out `dst` format from `fmt` param if specified or from `dst` filename.
> * fail if we can't guess it obviously
> * if different, then convert
> * call optimization

I have added a `allow_convert` option to optimize_image in `optimization.py` and exposed it in `__init__`. I think it'd be better than having two functions.

**Review by rgaudin — CHANGES_REQUESTED** [review]
That looks really nice now!

contributed a modified version of `optimize_image` that:
* don't assume source format from extension (we can't trust that)
* allows specifying the format to convert to and not convert or auto-select
* is simpler to understand

That required editing `format_for` to also work with Pillow. Also updated a test to match new signature.

Some minor changes in error management ; looking at presets now

File: src/zimscraperlib/image/optimization.py [hunk 1]
```diff
@@ -2,291 +2,299 @@
...
+        grayscale: Whether to convert image to grayscale (boolean)
+            values: True | False"""
+
+    ensure_matches(src, "PNG")
+
+    # use a temporary file as source as optimization is done destructively
+    tmp_path = get_temporary_copy(src)
+
+    # generate PNG task for optimize_images
+    # max_w and max_h is 0 because we have a better image resizing function in scraperlib already
```
[the remark this answers is not in the document]
Comment (satyamtg) [reply src/zimscraperlib/image/optimization.py:104]: Done

File: src/zimscraperlib/image/optimization.py [hunk 2]
```diff
@@ -2,291 +2,299 @@
...
+    if max_colors:
+        args += ["--colors", str(max_colors)]
+    if lossiness:
+        args += [f"--lossy={lossiness}"]
+    if no_extensions:
+        args += ["--no-extensions"]
+    if interlace:
+        args += ["--interlace"]
+    args += [str(src)]
+    logger.debug(nicer_args_join(args))
```
[the remark this answers is not in the document]
Comment (satyamtg) [reply src/zimscraperlib/image/optimization.py:261]: Removed it.

File: src/zimscraperlib/image/optimization.py [hunk 3]
```diff
@@ -2,291 +2,299 @@
...
+    with open(dst, "w") as out_file:
+        gifsicle = subprocess.run(args, stdout=out_file)
+    gifsicle.check_returncode()
```
[the remark this answers is not in the document]
Comment (satyamtg) [reply src/zimscraperlib/image/optimization.py:264]: Done

File: src/zimscraperlib/image/optimization.py [hunk 4]
```diff
@@ -2,291 +2,299 @@
...
+
+    ensure_matches(src, "WEBP")
+
+    webp_image = Image.open(src)
+    params = {
+        "lossless": lossless,
+        "quality": quality,
+        "method": method,
+    }
+    save_image(webp_image, dst, fmt="WEBP", **params)
```
[the remark this answers is not in the document]
Comment (satyamtg) [reply src/zimscraperlib/image/optimization.py:218]: Done

File: src/zimscraperlib/image/optimization.py [hunk 5]
```diff
@@ -2,291 +2,299 @@
...
+    shutil.copy(src, tmp_path)
+    return tmp_path
+
+
+def run_external_optimizer(
+    optimization_task: Task, tmp_path: pathlib.Path, dst: pathlib.Path
+) -> bool:
+    result = do_optimization(optimization_task)
+    if result.was_optimized:
+        shutil.copy(tmp_path, dst)
```
[the remark this answers is not in the document]
Comment (satyamtg) [reply src/zimscraperlib/image/optimization.py:54]: Done

**Review by rgaudin — APPROVED** [review]

**Pull request merged**